pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go

```diff
@@ -141,6 +150,248 @@ func getIndexIDForValidationForConstraint(b BuildCtx, tableID catid.DescID) (ret
+		for _, colName := range fkDef.FromCols {
+			colID := mustGetColumnIDFromColumnName(b, tbl.TableID, colName)
+			colHasOnUpdate := retrieveColumnOnUpdateExpressionElem(b, tbl.TableID, colID) != nil
+			if colHasOnUpdate {
+				panic(pgerror.Newf(
+					pgcode.InvalidTableDefinition,
+					"cannot specify a foreign key update action and an ON UPDATE"+
+						" expression on the same column",
+				))
+			}
+		}
+	}
+
+	// 2. If this FK has SET NULL action (ON UPDATE or ON DELETE) && any one of
+	// the originColumns is NOT NULL, then panic with error.
+	if fkDef.Actions.Delete == tree.SetNull || fkDef.Actions.Update == tree.SetNull {
+		for i, colName := range fkDef.FromCols {
+			colID := mustGetColumnIDFromColumnName(b, tbl.TableID, colName)
+			colIsNullable := mustRetrieveColumnTypeElem(b, tbl.TableID, colID).IsNullable
+			if !colIsNullable {
+				panic(pgerror.Newf(pgcode.InvalidForeignKey,
+					"cannot add a SET NULL cascading action on column %q which has a NOT NULL constraint", fromColsFRNames[i],
+				))
+			}
+		}
+	}
+
+	// 3. If this FK has SET DEFAULT action (ON UPDATE or ON DELETE) && any one of
+	// the originColumns does not have a default expression (which implies NULL will
+	// be used as the default) && that column is NOT NULL, then panic with error.
+	if fkDef.Actions.Delete == tree.SetDefault || fkDef.Actions.Update == tree.SetDefault {
+		for i, colName := range fkDef.FromCols {
+			colID := mustGetColumnIDFromColumnName(b, tbl.TableID, colName)
+			colHasDefault := retrieveColumnDefaultExpressionElem(b, tbl.TableID, colID) != nil
+			colIsNullable := mustRetrieveColumnTypeElem(b, tbl.TableID, colID).IsNullable
+			if !colHasDefault && !colIsNullable {
+				panic(pgerror.Newf(pgcode.InvalidForeignKey,
+					"cannot add a SET DEFAULT cascading action on column %q which has a "+
+						"NOT NULL constraint and a NULL default expression", fromColsFRNames[i],
+				))
+			}
+		}
+	}
+
+	// 4. Check whether each originColumns can be used for an outbound FK constraint, and
+	//    no duplicates exist in originColumns.
+	var originColIDs []catid.ColumnID
+	var originColSet catalog.TableColSet
+	for i, colName := range fkDef.FromCols {
+		colID := mustGetColumnIDFromColumnName(b, tbl.TableID, colName)
+		ensureColCanBeUsedInOutboundFK(b, tbl.TableID, colID)
+		if originColSet.Contains(colID) {
+			panic(pgerror.Newf(pgcode.InvalidForeignKey,
+				"foreign key contains duplicate column %q", fromColsFRNames[i]))
+		}
+		originColSet.Add(colID)
+		originColIDs = append(originColIDs, colID)
+	}
+
+	// 5. Resolve `t.(*ForeignKeyConstraintTableDef).Table` (i.e. referenced table name) and check whether it's in the same
+	//    database as the originTable (i.e. tbl). Cross database FK references is a deprecated feature and is in practice
+	//    no longer supported. We will return an error here directly.
+	referencedTableID := mustGetTableIDFromTableName(b, fkDef.Table)
```

Comment:
`mustGetTableIDFromTableName` requires `privilege.CREATE` to resolve the table, which seems odd for a table referenced in a FK. Can you double-check that this is the existing behaviour, please?